:memo: updated openapi documentation to include generation of client (#854)
it-at-m/refarch: docs/cross-cutting-concepts/openapi.md

These endpoints are normally used for client generation, validation or export of the documentation as this is a standardized format for any OpenAPI generator.

> ## Documenting endpoints

Reviewer: We should add information that one needs to add things like
```
@RequestMapping(value = "/theEntity", produces = MediaType.APPLICATION_JSON_VALUE)
@SecurityRequirement(name = OpenAPIDocumentationConfiguration.SECURITY_SCHEME_NAME)
```
to the controllers.

  return service.doSth(someId);
}
```

## Generating API specification

> The current API specification can be generated locally as a `.yaml` file using the [Springdoc Maven plugin](https://springdoc.org/#maven-plugin).

Reviewer: ```suggestion
An OpenAPI specification for the Spring API layer of the backend (e.g. controllers and DTOs) can be generated locally as a `.yaml` file using the [Springdoc Maven plugin](https://springdoc.org/#maven-plugin).
```

This is useful for different use-cases e.g.:

- OpenAPI-based client generation
- External documentation processing

The export is performed in the backend via the following command:

```shell
mvn springdoc-openapi:generate
```

> ::: info Information

Reviewer: I'd change this to a different color and mark it as Warning / Error.

Author: Running the application beforehand is purely informational - i dont see the point of using error or warning here.

> The backend component must be running during this process. The plugin fetches the OpenAPI definition from the running instance via its `http://localhost:8086/v3/api-docs` endpoint.

Reviewer: ```suggestion
The backend component must be running during this process, because the plugin fetches the OpenAPI specification via its `/v3/api-docs` endpoint.
```

:::

> The OpenAPI `.yaml` file will be saved into the backend's `api-spec` folder. The name is related to the Maven artifact itself (`{artifact-name}.yaml`).

Reviewer: ```suggestion
The OpenAPI `.yaml` file is saved in the `api-spec` folder of the backend. The file name is derived from the Maven artifact name (`{artifact-name}.yaml`).
```

> This file will be checked into the vcs, as it can be used inside the [following step](#generating-api-client-from-specification).

Reviewer: ```suggestion
This file should be checked into VCS when used for the [API client generation](#generating-api-client-from-specification).
```

::: danger IMPORTANT

> Changing the output file location or name via plugin configuration is highly discouraged, as other processes (e.g., client generation) might depend on the default.

Reviewer: ```suggestion
Changing the file name or output location in the plugin configuration is highly discouraged, as (out of the box) other features (e.g. client generation) depend on the default plugin configuration.
```

> Do not edit the generated OpenAPI specification file manually, regenerate it from the backend sources instead.

Reviewer: ```suggestion
The generated OpenAPI specification must not be edited manually. Instead, it should be regenerated from backend sources using the Maven plugin.
```

:::

## Generating API client from specification

The generator is configured via the `openapitools.json` file in the frontend / webcomponent project, which references the exported OpenAPI specification from the backend.
The generated API specification from the [previous step](#generating-api-specification) can be used to generate a client for the API inside the [frontend or webcomponent template](../templates/getting-started.md#frontend-web-components).

> For this the `@openapitools/openapi-generator-cli` dependency is used, which can be started via the `pre-build` script. This script will be automatically called while using the `build` or `lint` script.

Reviewer: ```suggestion

The generated API specification from the [previous step](#generating-api-specification) can be used to generate API clients for the [frontend or webcomponent template](../templates/getting-started.md#frontend-web-components).
This is achieved using the [`@openapitools/openapi-generator-cli`](https://openapi-generator.tech/docs/usage). `openapitools.json` configures the generator, while this file also references the exported OpenAPI specification from the backend. The generation step can be executed manually using `npm run pre-build` and is also automatically executed when running `build` or `lint`.
```

Question: Shouldn't it also be executed when running `fix`?

Author: We do not execute `fix` inside the pipeline execution. This script should be kept as fast as possible.

Reviewer: Fine for me, however I guess when developers only run `fix` locally, they might get false positive issues when the code was not generated yet.

```json:line-numbers=10
  "scripts": {
    "dev": "vite",
    "test": "vitest run",
    "build": "npm run pre-build && vue-tsc --build --noCheck && vite build && npm run post-build",
    "pre-build": "openapi-generator-cli generate", // [!code focus]
    "post-build": "cyclonedx-npm --omit dev --omit optional --output-reproducible -o dist/application.cdx.json",
    "lint": "npm run pre-build && prettier . --check --cache && eslint . --cache --concurrency auto && vue-tsc --build --noEmit",
    "fix": "prettier . --write --cache && eslint . --fix --cache --concurrency auto"
  },

> ```

Reviewer: I don't think this is necessary here, as the job is already mentioned above. Also this block is tedious to sync if we update any of the npm scripts in the templates.

Author: It looks very nice tho

Author: I could see myself being persuaded to only display the pre-build script

### Typical workflow

1. Generate / Update the OpenAPI specification in the backend (see [Generating API specification](#generating-api-specification))
2. Run `npm run build` or `npm run lint` inside the frontend or webcomponent project

> The configuration inside `openapitools.json` normally does not need any modification as its independent of the specification's filename.

Reviewer: I'd add information that this file only has to be modified, when multiple OpenAPI specs (e.g. multiple backends are used)

> The configuration inside `openapitools.json` normally does not need any modification as its independent of the specification's filename.

Reviewer: ```suggestion
The configuration inside `openapitools.json` normally does not require any modification as its independent of the specification's filename.
```

### Generated folder structure

The generated client will be placed inside this folder: `src/api/generated/<backend-artifact>`.

> The naming in `<backend-artifact>` differs from project to project and results from the specifications filename.

Reviewer: ```suggestion
The generated API client is placed inside the folder `src/api/generated/<spec-filename>`.
The naming in `<backend-artifact>` differs from project to project and results from the specifications filename.
```

> Each controller has its own API class (`src/api/generated/<backend-artifact>/apis/*`).

Reviewer: ```suggestion
For each Spring controller a dedicated API class (`src/api/generated/<spec-filename>/apis/*`) is generated.
```

> These should be accessed via the provided `ApiFactory.ts` inside of the components.

Reviewer: ```suggestion
Those API client classes should be used via the `ApiFactory.ts` helper and its `getInstance` method.
```

Maybe also add an example call here?

> All DTOs necessary for requesting and receiving information can be found inside the `models` folder, together with all restriction set to attributes inside the backend (e.g. length of a string).

Reviewer: ```suggestion
All DTOs for requests and responses are stored inside the `models` folder. Field constraints of the Java DTOs (e.g. `@NotNull` or `@Size`) are respected in the generated TypeScript code as well.
```

Author: I like that

### First usage

After installing the new dependency, we recommend to execute the following script before first usage.
Otherwise, the npm-script might still throw an error upon execution as it could not download the `openapi-generator-cli-X.X.X.jar` while using a proxy.

> This `.jar` is mandatory, as it contains the generator itself. The node dependency is just a wrapper around this `.jar`.

Reviewer: ```suggestion
This `.jar` file is mandatory, as it contains the generator logic. The NPM dependency is just a wrapper around this `.jar` file.
```

::: code-group

```shell [shell]
./download-openapi-jar.sh
```

```shell [bash]
./download-openapi-jar.bat
```

This shell script automatically extracts the needed version from the `openapitools.json`, downloads the needed `.jar` and places into your `node_modules` folder.

:::

::: info Information
This script needs to be executed each time you update the version inside your `openapitools.json`.

> :::

Reviewer: I'd move this whole section into "Generating API client from specification" as a simple info box that describes running the script is likely necessary when using it behind a proxy.
The purpose of the script (downloading the `.jar` file containing the generator logic, and that NPM dep is just a wrapper, should still be mentioned)

Author: Ok - this works only with a small workaround but we can try

### Usage of frontend without backend

> If the frontend or web component is used without a backend (and therefore without an API specification), the scripts previously mentioned scripts for generting the client inside the `package.json` need some adjustment.

Reviewer: ```suggestion
If the frontend or web component is used without a backend (and therefore without an API specification), the previously mentioned script generating the client inside the `package.json` need to be adjusted. The `pre-build` script must be deleted and its reference in the `build` and `lint` scripts must be removed as well.
```

Author: Currently the pre-build script only contains the api-generator but it may contain some more script in the future Therefore i would advise against the removal of the whole script

This can be done via removing the `openapi-generator-cli` call inside the `pre-build` step.

```json:line-numbers=10
  "scripts": {
    "dev": "vite",
    "test": "vitest run",
    "build": "npm run pre-build && vue-tsc --build --noCheck && vite build && npm run post-build",
    "pre-build": "openapi-generator-cli generate", // [!code --]
    "pre-build": "", // [!code ++]
    "post-build": "cyclonedx-npm --omit dev --omit optional --output-reproducible -o dist/application.cdx.json",
    "lint": "npm run pre-build && prettier . --check --cache && eslint . --cache --concurrency auto && vue-tsc --build --noEmit",
    "fix": "prettier . --write --cache && eslint . --fix --cache --concurrency auto"
  },

> ```

Reviewer: Actually this is only relevant when not using API generation at all, so I think we should move it somewhere to the top as a information inside a info box.

Author: I dont like to overuse these info boxes. Furthermore this information is only needed in edge cases. Therefore appending it should be fine.

  return service.doSth(someId);
}
```

## Generating API specification

An OpenAPI specification for the backend API layer (e.g. controllers and DTOs) can be generated locally as a `.yaml` file using the [Springdoc Maven plugin](https://springdoc.org/#maven-plugin).

This is useful for different use-cases e.g.:

- OpenAPI-based client generation
- External documentation processing

The export is performed in the backend via the following command:

```shell
mvn springdoc-openapi:generate
```

::: info Information
The backend component must be running during this process, because the plugin fetches the OpenAPI specification via its `/v3/api-docs` endpoint.
:::

Further notice, every controller should be annotated as follows:

```java
@RequestMapping(..., produces = MediaType.APPLICATION_JSON_VALUE)
@SecurityRequirement(name = OpenAPIDocumentationConfiguration.SECURITY_SCHEME_NAME)
```

On the `@RequestMapping`, the `produces` media type is explicitly defined to ensure that the generated specification uses the correct media type (`application/json`) instead of the generic `*/*`.

> The `@SecurityRequirement` annotation indicates that this controller is secured using the global `SecurityConfiguration`. Adjust this as required based on your project’s security setup.

Reviewer: ```suggestion
The `@SecurityRequirement` annotation indicates that this controller is secured using the global `SecurityConfiguration`. Adjust this as required based on the security setup of the project.
```

Neither of these annotations affects OpenAPI client generation. They only improve the accuracy and completeness of the generated specification.

> The OpenAPI `.yaml` file is saved in the backend's `api-spec` folder.

Reviewer: ```suggestion
The OpenAPI `.yaml` file is saved in the `api-spec` folder of the backend.
```

The filename is derived from the Maven artifact defined inside the `pom.xml`.
By default, this file is and should be kept under version control as it may be used inside the [following step](#generating-api-client-from-specification).

::: danger IMPORTANT

> Changing the name or location via plugin configuration is highly discouraged, as other features (e.g., client generation) might depend on the default configuration.

Reviewer: ```suggestion
Changing the name or location via plugin configuration is highly discouraged, as other features (e.g. client generation) might depend on the default configuration.
```

> The generated OpenAPI specification must not be edited manually as any generated file. Instead, regenerate it from backend sources using the previously described Maven plugin.

Reviewer: ```suggestion
As any generated file, the OpenAPI specification must not be edited. Instead, it should be regenerated from backend sources using the previously described Maven plugin.
```

:::

## Generating API client from specification

The generator is configured via the `openapitools.json` file in the frontend / webcomponent project, which references the exported OpenAPI specification from the backend.

> The generated API specification from the [previous step](#generating-api-specification) can be used to generate Api clients for the [frontend or webcomponent template](../templates/getting-started.md#frontend-web-components).

Reviewer: ```suggestion
The generated API specification from the [previous step](#generating-api-specification) can be used to generate API clients for the [frontend or webcomponent template](../templates/getting-started.md#frontend-web-components).
```

Author: Replaced all instances

> For this purpose the [`@openapitools/openapi-generator-cli`](https://openapi-generator.tech/docs/usage) dependency is added, which can be executed manually via the `pre-build` script. This script will be automatically called while using the `build` or `lint` script.

Reviewer: ```suggestion
For this purpose the [`@openapitools/openapi-generator-cli`](https://openapi-generator.tech/docs/usage) dependency is used. It runs automatically when using the `build` and `lint` scripts or manually via the `pre-build` script.
```

```json:line-numbers=10
  "scripts": {
    ...
    "pre-build": "openapi-generator-cli generate", // [!code focus]
    ...
  },
```

::: info Usage behind a proxy

> When using a proxy, it might be necessary to run the custom script we provide after installing the dependency.

Reviewer: ```suggestion
When using a proxy, it might be necessary to run the provided custom script after installing the dependency.
```

> This scripts downloads the required `.jar` file that includes the generator logic.

Reviewer: ```suggestion
This script downloads additional required artifacts that includes the generator logic. It automatically extracts the relevant version from the `openapitools.json`, downloads the required `.jar` file and places into the project's `node_modules` folder.
```

The NPM dependency only provides a wrapper around this `.jar`.

::: code-group

```shell [shell]
./download-openapi-jar.sh
```

```shell [bash]
./download-openapi-jar.bat
```

> This shell script automatically extracts the relevant version from the `openapitools.json`, downloads the needed `.jar` and places into the project's `node_modules` folder.

Reviewer: ```suggestion
```

> This script needs to be executed each time the version inside the `openapitools.json` is updated.

Reviewer: ```suggestion
The script needs to be executed again when the generator version in `openapitools.json` is changed.
```

:::

### Typical workflow

1. Generate / Update the OpenAPI specification in the backend (see [Generating API specification](#generating-api-specification))
2. Run `npm run build` or `npm run lint` inside the frontend or webcomponent project

If multiple clients (e.g. based on different specifications) are to be generated, the configuration in `openapitools.json` must be adapted accordingly.
Otherwise, no modifications are necessary, as the default configuration already handles standard use cases (e.g. independence from the specification filename).

### Generated folder structure

After generating, the API client is placed inside the folder `src/api/generated/<specification-filename>`.

> The folders name is derived from the specifications' filename.

Reviewer: ```suggestion
The folder name is derived from the filename of the specification.
```

> For each Spring controller a dedicated API class (`src/api/generated/<specification-filename>/apis/*`) is generated and should be used via the provided `ApiFactory.ts` and its `getInstance` function.

Reviewer: ```suggestion
For each Spring controller a dedicated API class (`src/api/generated/<specification-filename>/apis/*`) is generated. Those should be used via the provided `ApiFactory.ts` and its `getInstance` function.
```

Maybe also add an example snippet here?

> All DTOs for requests and response are generated inside the `models` folder.

Reviewer: ```suggestion
All DTOs for requests and responses are generated inside the `models` folder.
```

Field constraints of the Java DTOs (e.g. `@NotNull` or `@Size`) are respected in the generated TypeScript code as well.

### Usage of frontend without backend

> If the frontend or web component is used without a backend (and therefore without an API specification), the previously mentioned scripts for generating the client inside the `package.json` needs to be adjusted.

Reviewer: ```suggestion
If the frontend or web component is used without a backend (and therefore without an API specification), the previously mentioned script for generating the client inside the `package.json` needs to be adjusted.
```

> This can be done via removing the `openapi-generator-cli` call inside the `pre-build` step.

Reviewer: ```suggestion
Specifically, the call of `openapi-generator-cli` inside the `pre-build` step has to be removed.
```

```json:line-numbers=10
  "scripts": {
    ...
    "pre-build": "openapi-generator-cli generate", // [!code --][!code focus]
    "pre-build": "", // [!code ++] [!code focus]
    ...
  },
```

> No further adjustment is needed.

Reviewer: ```suggestion
No further adjustment is required.
```

  return service.doSth(someId);
}
```

## Generating API specification

An OpenAPI specification for the backend API layer (e.g. controllers and DTOs) can be generated locally as a `.yaml` file using the [Springdoc Maven plugin](https://springdoc.org/#maven-plugin).

This is useful for different use-cases e.g.:

- OpenAPI-based client generation
- External documentation processing

The export is performed in the backend via the following command:

```shell
mvn springdoc-openapi:generate
```

::: info Information
The backend component must be running during this process, because the plugin fetches the OpenAPI specification via its `/v3/api-docs` endpoint.
:::

Further notice, every controller should be annotated as follows:

```java

> @RequestMapping(value = "/theEntity", produces = MediaType.APPLICATION_JSON_VALUE)

Author: ```suggestion
@RequestMapping(..., produces = MediaType.APPLICATION_JSON_VALUE)
```

  return service.doSth(someId);
}
```

## Generating API specification

> An OpenApi specification for the backend API layer (e.g. controllers and DTOs) can be generated locally as a `.yaml` file using the [Springdoc Maven plugin](https://springdoc.org/#maven-plugin).

Reviewer: Please `OpenApi` with `OpenAPI` everywhere.

This is useful for different use-cases e.g.:

- OpenAPI-based client generation
- External documentation processing

The export is performed in the backend via the following command:

```shell
mvn springdoc-openapi:generate
```

::: info Information
The backend component must be running during this process, because the plugin fetches the OpenAPI specification via its `/v3/api-docs` endpoint.
:::

Further notice, every controller should be annotated as follows:

```java

> @RequestMapping(value = "/theEntity", produces = MediaType.APPLICATION_JSON_VALUE)

Reviewer: Maybe only include the relevant part as others might have request mappings directly on controller methods? Also this prevents copy paste mistakes.

```suggestion
@RequestMapping(produces = MediaType.APPLICATION_JSON_VALUE)
```

Author: Mhh i dont like this as it may mislead the developer to remove the value attribute. `...` would be fine with me

@SecurityRequirement(name = OpenAPIDocumentationConfiguration.SECURITY_SCHEME_NAME)
```

On the `@RequestMapping`, the `produces` media type is explicitly defined to ensure that the generated specification uses the correct media type (`application/json`) instead of the generic `*/*`.

> The `@SecurityRequirement` annotation indicates that this controller is secured using the global `WebAuthConfiguration`. Adjust this as required based on your project’s security setup.

Reviewer: What is `WebAuthConfiguration`? Shouldn't this be `OpenAPIDocumentationConfiguration`?

Author: No - but i adjusted the naming to represent the class name inside the backend

Neither of these annotations affects OpenAPI client generation. They only improve the accuracy and completeness of the generated specification.

> The OpenAPI `.yaml` file will be saved in the backend's `api-spec` folder.

Reviewer: ```suggestion
The OpenAPI `.yaml` file is generated in the backend's `api-spec` folder.
```

The filename is derived from the Maven artifact defined inside the `pom.xml`.

> By default, this file is and should be versioned controlled as it may be used inside the [following step](#generating-api-client-from-specification).

Reviewer: ```suggestion
By default, this file is and should be version-controlled as it may be used inside the [following step](#generating-api-client-from-specification).
```

::: danger IMPORTANT

> Changing the name or location via plugin configuration is highly discouraged, as other features (e.g., client generation) might depend on the default / RefArch plugin configuration.

Reviewer: ```suggestion
Changing the name or location via plugin configuration is highly discouraged, as other features (e.g., client generation) might depend on the default configuration.
```

RefArch plugin configuration is the default, so I'd omit it.

> The generated OpenApi specification must not be edited manually as any generated file. Instead, regenerate it from backend sources using the previously described Maven plugin.

Reviewer: ```suggestion
As any generated file, the OpenAPI specification must not be edited manually. Instead, regenerate it from backend sources using the previously described Maven plugin.
```

:::

## Generating API client from specification

The generator is configured via the `openapitools.json` file in the frontend / webcomponent project, which references the exported OpenAPI specification from the backend.
The generated API specification from the [previous step](#generating-api-specification) can be used to generate Api clients for the [frontend or webcomponent template](../templates/getting-started.md#frontend-web-components).

> For this purpose the [`@openapitools/openapi-generator-cli`](https://openapi-generator.tech/docs/usage) dependency was added.

Reviewer: Please just present tense. `is used` vs `was added`

> For this the `@openapitools/openapi-generator-cli` dependency is used, which can be started via the `pre-build` script. This script will be automatically called while using the `build` or `lint` script.

Reviewer: Both sentences start almost the same. "For this ... the `@openapitools/openapi-generator-cli` dependency". Please rewrite..

```json:line-numbers=10
  "scripts": {
    ...
    "pre-build": "openapi-generator-cli generate", // [!code focus]
    ...
  },
```

> ::: info First usage

Reviewer: ```suggestion
::: info Usage behind a proxy
```

After installing the new dependency, we recommend to execute the following script before first usage.
Otherwise, the npm-script might still throw an error upon execution as it could not download the `openapi-generator-cli-X.X.X.jar` while using a proxy.
This `.jar` file is required, as it contains the actual generator.

> The npm dependency only provides a wrapper around this `.jar`.

Reviewer: ```suggestion
When using a proxy, it might be necessary to run the custom script we provide after installing the dependency.
This scripts downloads the required `.jar` file that includes the generator logic, while the NPM itself only acts as a wrapper.
```

::: code-group

```shell [shell]
./download-openapi-jar.sh
```

```shell [bash]
./download-openapi-jar.bat
```

> This shell script automatically extracts the needed version from the `openapitools.json`, downloads the needed `.jar` and places into your `node_modules` folder.

Reviewer: ```suggestion
This shell script automatically extracts the relevant version from the `openapitools.json`, downloads the required `.jar` and places into the local`node_modules` folder.
```

> This script needs to be executed each time you update the version inside your `openapitools.json`.

Reviewer: ```suggestion
This script needs to be executed each time the version inside `openapitools.json` is updated.
```

Use impersonal form.

:::

### Typical workflow

1. Generate / Update the OpenAPI specification in the backend (see [Generating API specification](#generating-api-specification))
2. Run `npm run build` or `npm run lint` inside the frontend or webcomponent project

If multiple clients (e.g. based on different specifications) are to be generated, the configuration in `openapitools.json` must be adapted accordingly.
Otherwise, no modifications are necessary, as the default configuration already handles standard use cases (e.g. independence from the specification filename).

### Generated folder structure

After generating, the API client is placed inside the folder `src/api/generated/<specification-filename>`.
The folders name is derived from the specifications' filename.

> For each Spring controller a dedicated API class (`src/api/generated/<backend-artifact>/apis/*`) is generated and should be used via the provided `ApiFactory.ts` and its `getInstance` function.

Reviewer: ```suggestion
For each Spring controller a dedicated API class (`src/api/generated/<specification-filename>/apis/*`) is generated and should be used via the provided `ApiFactory.ts` and its `getInstance` function.
```
